src/node_http2.cc

```diff
@@ -0,0 +1,1326 @@
+#include "node.h"
+#include "node_buffer.h"
+#include "node_http2.h"
+
+namespace node {
+
+using v8::ArrayBuffer;
+using v8::Boolean;
+using v8::Context;
+using v8::Function;
+using v8::Integer;
+using v8::Undefined;
+
+namespace http2 {
+
+enum Http2SettingsIndex {
+  IDX_SETTINGS_HEADER_TABLE_SIZE,
+  IDX_SETTINGS_ENABLE_PUSH,
+  IDX_SETTINGS_INITIAL_WINDOW_SIZE,
+  IDX_SETTINGS_MAX_FRAME_SIZE,
+  IDX_SETTINGS_MAX_CONCURRENT_STREAMS,
+  IDX_SETTINGS_MAX_HEADER_LIST_SIZE,
+  IDX_SETTINGS_COUNT
+};
+
+enum Http2SessionStateIndex {
+  IDX_SESSION_STATE_EFFECTIVE_LOCAL_WINDOW_SIZE,
+  IDX_SESSION_STATE_EFFECTIVE_RECV_DATA_LENGTH,
+  IDX_SESSION_STATE_NEXT_STREAM_ID,
+  IDX_SESSION_STATE_LOCAL_WINDOW_SIZE,
+  IDX_SESSION_STATE_LAST_PROC_STREAM_ID,
+  IDX_SESSION_STATE_REMOTE_WINDOW_SIZE,
+  IDX_SESSION_STATE_OUTBOUND_QUEUE_SIZE,
+  IDX_SESSION_STATE_HD_DEFLATE_DYNAMIC_TABLE_SIZE,
+  IDX_SESSION_STATE_HD_INFLATE_DYNAMIC_TABLE_SIZE,
+  IDX_SESSION_STATE_COUNT
+};
+
+enum Http2StreamStateIndex {
+  IDX_STREAM_STATE,
+  IDX_STREAM_STATE_WEIGHT,
+  IDX_STREAM_STATE_SUM_DEPENDENCY_WEIGHT,
+  IDX_STREAM_STATE_LOCAL_CLOSE,
+  IDX_STREAM_STATE_REMOTE_CLOSE,
+  IDX_STREAM_STATE_LOCAL_WINDOW_SIZE,
+  IDX_STREAM_STATE_COUNT
+};
+
+enum Http2OptionsIndex {
+  IDX_OPTIONS_MAX_DEFLATE_DYNAMIC_TABLE_SIZE,
+  IDX_OPTIONS_MAX_RESERVED_REMOTE_STREAMS,
+  IDX_OPTIONS_MAX_SEND_HEADER_BLOCK_LENGTH,
+  IDX_OPTIONS_PEER_MAX_CONCURRENT_STREAMS,
+  IDX_OPTIONS_PADDING_STRATEGY,
+  IDX_OPTIONS_FLAGS
+};
+
+Http2Options::Http2Options(Environment* env) {
+  nghttp2_option_new(&options_);
+
+  uint32_t* buffer = env->http2_options_buffer();
+  uint32_t flags = buffer[IDX_OPTIONS_FLAGS];
+
+  if ((flags & (1 << IDX_OPTIONS_MAX_DEFLATE_DYNAMIC_TABLE_SIZE)) ==
+      (1 << IDX_OPTIONS_MAX_DEFLATE_DYNAMIC_TABLE_SIZE)) {
```

Comment:
Btw, I’d find these (in C++ and in JS) more readable if it were just
`if (flags & (1 << IDX_OPTIONS_MAX_DEFLATE_DYNAMIC_TABLE_SIZE)) { … }` etc.